src/arm/windows/init_by_logical_sys_info.c

```diff
@@ -0,0 +1,842 @@
+
+	read_cores_for_processors(
+		processors, nr_of_processors,
+		global_proc_index_per_group, cores,
+		chip_info);
+
+	/* 5. Now that we read out everything from the system we can, fill the package, cluster
+	 *    and core structures respectively.
+	 */
+	 connect_packages_cores_clusters_by_processors(
+		processors, nr_of_processors,
+		packages, nr_of_packages,
+		clusters,
+		cores, nr_of_cores,
+		chip_info,
+		vendor);
+
+	/* 6. Count and store uarchs of cores, assuming same uarchs are neighbors */
+	enum cpuinfo_uarch prev_uarch = cpuinfo_uarch_unknown;
+	for (uint32_t i = 0; i < nr_of_cores; i++) {
+		if (prev_uarch != cores[i].uarch) {
+			nr_of_uarchs++;
+			prev_uarch = cores[i].uarch;
+		}
+	}
+	uarchs = HeapAlloc(heap, HEAP_ZERO_MEMORY, nr_of_uarchs * sizeof(struct cpuinfo_uarch_info));
+	if (uarchs == NULL) {
+		cpuinfo_log_error("failed to allocate %zu bytes for descriptions of %"PRIu32" uarchs",
+			nr_of_uarchs * sizeof(struct cpuinfo_uarch_info), nr_of_uarchs);
+		goto clean_up;
+	}
+	prev_uarch = cpuinfo_uarch_unknown;
+	for (uint32_t i = 0, uarch_counter = 0; i < nr_of_cores; i++) {
+		if (prev_uarch != cores[i].uarch) {
+			prev_uarch = cores[i].uarch;
+			uarchs[uarch_counter].uarch = cores[i].uarch;
+			uarchs[uarch_counter].core_count = 1;
+			uarchs[uarch_counter].processor_count = cores[i].processor_count;
+			uarch_counter++;
+		} else if (prev_uarch != cpuinfo_uarch_unknown) {
+			uarchs[uarch_counter].core_count++;
+			uarchs[uarch_counter].processor_count += cores[i].processor_count;
+		}
+	}
+
+	/* 7. Commit changes */
+	cpuinfo_processors = processors;
+	cpuinfo_packages = packages;
+	cpuinfo_clusters = clusters;
+	cpuinfo_cores = cores;
+	cpuinfo_uarchs = uarchs;
+
+	cpuinfo_processors_count = nr_of_processors;
+	cpuinfo_packages_count = nr_of_packages;
+	cpuinfo_clusters_count = nr_of_cores;
+	cpuinfo_cores_count = nr_of_cores;
+	cpuinfo_uarchs_count = nr_of_uarchs;
+
+	for (uint32_t i = 0; i < MAX_NR_OF_CACHES; i++) {
+		cpuinfo_cache_count[i] = numbers_of_caches[i];
+	}
+	cpuinfo_cache[cpuinfo_cache_level_1i] = caches;
+	cpuinfo_cache[cpuinfo_cache_level_1d] = cpuinfo_cache[cpuinfo_cache_level_1i] + cpuinfo_cache_count[cpuinfo_cache_level_1i];
+	cpuinfo_cache[cpuinfo_cache_level_2]  = cpuinfo_cache[cpuinfo_cache_level_1d] + cpuinfo_cache_count[cpuinfo_cache_level_1d];
+	cpuinfo_cache[cpuinfo_cache_level_3]  = cpuinfo_cache[cpuinfo_cache_level_2]  + cpuinfo_cache_count[cpuinfo_cache_level_2];
+	cpuinfo_cache[cpuinfo_cache_level_4]  = cpuinfo_cache[cpuinfo_cache_level_3]  + cpuinfo_cache_count[cpuinfo_cache_level_3];
+	cpuinfo_max_cache_size = cpuinfo_compute_max_cache_size(&processors[0]);
+
+	result = true;
+	MemoryBarrier();
+
+	processors = NULL;
+	packages = NULL;
+	clusters = NULL;
+	cores = NULL;
+	caches = NULL;
+	uarchs = NULL;
+
+clean_up:
+	if (global_proc_index_per_group != NULL) {
+		HeapFree(heap, 0, global_proc_index_per_group);

```

Comment:
its safe to call HeapFree() with a null ptr for the 3rd param. 
Drop the NULL check and just call it.
This is a style thing, but I would always zero the variable after calling free just to show its not usable past that point.